ode/unit_test/Test_ODE_Newton.hpp

```diff
@@ -0,0 +1,542 @@
+    auto local_x = Kokkos::subview(
+        x, Kokkos::pair(my_nls.neqs * idx, my_nls.neqs * (idx + 1)));
+    auto local_rhs = Kokkos::subview(
+        rhs, Kokkos::pair(my_nls.neqs * idx, my_nls.neqs * (idx + 1)));
+    auto local_update = Kokkos::subview(
+        update, Kokkos::pair(my_nls.neqs * idx, my_nls.neqs * (idx + 1)));
+    auto local_J = Kokkos::subview(
+        J, Kokkos::pair(my_nls.neqs * idx, my_nls.neqs * (idx + 1)),
+        Kokkos::ALL());
+    auto local_tmp = Kokkos::subview(
+        tmp, Kokkos::pair(my_nls.neqs * idx, my_nls.neqs * (idx + 1)),
+        Kokkos::ALL());
+
+    // Run Newton nonlinear solver
+    status(idx) = KokkosODE::Experimental::Newton::Solve(
+        my_nls, params, local_J, local_tmp, local_x, local_rhs, local_update);
+  }
+};
+
+template <class system_type, class execution_space, class scalar_type>
+void run_newton_test(const system_type& mySys,
+                     KokkosODE::Experimental::Newton_params& params,
+                     const scalar_type* const initial_val,
+                     const scalar_type* const solution) {
+  using newton_solver_status = KokkosODE::Experimental::newton_solver_status;
+  using vec_type = typename Kokkos::View<scalar_type*, execution_space>;
+  using mat_type = typename Kokkos::View<scalar_type**, execution_space>;
+
+  Kokkos::View<newton_solver_status*, execution_space> status("Newton status",
+                                                              1);
+
+  vec_type x("solution vector", mySys.neqs),
+      rhs("right hand side vector", mySys.neqs);
+  auto x_h = Kokkos::create_mirror_view(x);
+  auto r_h = Kokkos::create_mirror_view(rhs);
+
+  vec_type update("update", mySys.neqs);
+  mat_type J("jacobian", mySys.neqs, mySys.neqs),
+      tmp("temp mem", mySys.neqs, mySys.neqs + 4);
+
+  // Initial values
+  for (int eqIdx = 0; eqIdx < mySys.neqs; ++eqIdx) {
+    x_h(eqIdx) = initial_val[eqIdx];
+  }
+  Kokkos::deep_copy(x, x_h);
+
+  Kokkos::RangePolicy<execution_space> my_policy(0, 1);
+  NewtonSolve_wrapper solve_wrapper(mySys, params, x, rhs, update, J, tmp,
+                                    status);
+
+  Kokkos::parallel_for(my_policy, solve_wrapper);
+
+  auto status_h = Kokkos::create_mirror_view(status);
+  Kokkos::deep_copy(status_h, status);
+  EXPECT_TRUE(status_h(0) == newton_solver_status::NLS_SUCCESS);
+
+  Kokkos::deep_copy(x_h, x);
+  Kokkos::deep_copy(r_h, rhs);
+#ifdef HAVE_KOKKOSKERNELS_DEBUG
+  std::cout << "Non-linear problem solution and residual:" << std::endl;
+  std::cout << "  [(";
+  for (int eqIdx = 0; eqIdx < mySys.neqs; ++eqIdx) {
+    std::cout << " " << x_h(eqIdx);
+  }
+  std::cout << " ), " << KokkosBlas::serial_nrm2(rhs) << ", (";
+  for (int eqIdx = 0; eqIdx < mySys.neqs; ++eqIdx) {
+    std::cout << " "
+              << Kokkos::abs(x_h(eqIdx) - solution[eqIdx]) /
+                     Kokkos::abs(solution[eqIdx]);
+  }
+  std::cout << " )]" << std::endl;
+#else
+  (void)solution;
```

Comment:
Preferably there would at least be a loose check that ``solution`` is close to ``x_h``? Assuming you know which solution the solver will find in cases of there being multiple true solutions.